### ISSUE TITLE
Test plan for pattern matching generic type parameters

### ISSUE THREAD

**Issue by agocke** [dotnet/roslyn#19038]
This is my work-in-progress issue for a test plan for pattern matching generic parameters (https://github.com/dotnet/roslyn/issues/16195).

There should be tests for the following "golden path" scenarios:

* [X] Matching anything against an unconstrained generic parameter succeeds/fails at runtime [@gafter: See tests `TestMatchWithTypeParameter_01` and `TestMatchWithTypeParameter_02` in `CodeGen/SwitchTests.cs`]
* [X] Matching a class type against a type parameter constrained to that class type succeeds/fails at runtime
   - No compile-time diagnostics expected
* [x] Matching against a generic parameter constrained to class/struct succeeds/fails [@gafter see `TestMatchWithTypeParameter_03` and `TestMatchWithTypeParameter_04`]
    - Note: Spec currently suggests that a conversion always exists, and thus even when the check
      could be known to fail at compile time, it seems like no diagnostic is produced. Is this acceptable? [@gafter: We had to select specific criteria for what is detected and what is allowed to pass. We use the same criteria as we do for warning of an always `false` result for the existing `is` operator.]
  * [x] Verify definite assignment agrees
     - Note: If we can refute the match ahead of time should this be considered unreachable code? [@gafter: The subsumption tests confirm this. To pick an arbitrary one, see `TypeParameterSubsumption05`]
* [x] Matching a reference type against a generic parameter constrained to an interface [@gafter: adding a test for this exposed two bugs. See `TestMatchWithTypeParameter_05`]
* [x] Matching a value type against a generic parameter constrained to an interface [@gafter: see `TestMatchWithTypeParameter_06`] [@gafter testing this with the value type being `System.ValueTuple<int,int>` exposed more bugs; fixed with new tests including `ExplicitTupleConversion_Crash`.]
* [x] Matching against a generic parameter constrained with a `new()` constraint [@gafter: see `TestMatchWithTypeParameter_07`]

Edit: Subsumption seems to be spec'd and work OK with certain matching cases right now, e.g.

```csharp
void M<T, U>(T t, U u) where T : U
{
   switch((object)"test")
   {
     case U uu:
       break;
     case T tt: // Produces a diagnostic about subsumption/unreachability
       break;
}
```

* [x] We should have a similar test for matching without object. [@gafter: see `TypeParameterSubsumption05`]
* [x] Matching against `T` where `T` is substituted as `dynamic`. Always succeeds? (except for `null`) [@gafter see `GenericDynamicIsObject`]
* [x] Matching where the subject is `Nullable<T>` [@gafter: see `MatchNullableTypeParameter`]

---

* [x] "Recursive" generic [@gafter: see `MatchRecursiveGenerics`]
   * ```csharp
       using System;
       class Packet { } 
       class Packet<U> : Packet {} 
       public class C {
              void M<T>() where T : Packet {
                   T p = default;
                   if (p is Packet<T> p1) {
                   }    
              }
       }
       ```
* [x] Matching with the subject as restricted types, pointer types, `ref` types [@gafter for the first two, see `MatchRestrictedTypes`. We don't have a concept of `ref` types, so I don't know what you're suggesting for that. It is possible that `MatchRestrictedTypes` tests what you're suggesting.]

**Comment by agocke** [comment]
cc @gafter

**Comment by gafter** [comment]
Subsumption is a **very** weak area of the specification right now. 

The rule for subsumption of one type pattern `T t` by another type pattern `U u` (for any types T and U) is that the former is subsumed by the latter if there is an implicit predefined conversion from T to U. That applies in the case that both are type parameters and there is a constraint `where T : U`. There should be and probably isn't a test for that, but it would look something like this:

``` c#
    static void M<T, U>(object o) where T : U
    {
        switch (o)
        {
            case U u:
            case T t: // error: subsumed
                break;
        }
    }
```

**Comment by agocke** [comment]
@gafter I think this is ready to start working on. Most of the cases I listed were places where we could theoretically refute the match at compile time. It's not clear to me what we want to do in this case. It appears, from the `is` behavior, that we do not attempt to refute anything at compile time.

Maybe this is acceptable, but since it would be a breaking change to change this later, it's something we should encode in our tests.

**Comment by jaredpar** [comment]
@agocke what work is remaining here?

**Comment by agocke** [comment]
@jaredpar I've looked through the comments @gafter left. I think everything has been addressed, so I will mark this as completed.
